serialization.go

```diff
@@ -248,60 +248,104 @@ func IsPreIssuer(issuer *x509.Certificate) bool {
+// TLS marshaling.
+type UnmarshaledLeafEntry struct {
+	// Leaf is a MerkleTreeLeaf corresponding to LeafEntry.LeafInput.
+	Leaf MerkleTreeLeaf
+	// Cert is either certificate or pre-certificate.
+	Cert ASN1Cert
+	// Chain is the issuing certificate chain, starting with the issuer of Cert.
+	Chain []ASN1Cert
+	// IsPrecert determines what the Cert field contains. It corresponds to
+	// Leaf.TimestampedEntry.EntryType.
+	IsPrecert bool
+}
+
+// UnmarshalLeafEntry TLS-unmarshals the passed in LeafEntry, and stores the
+// result in the passed in UnmarshaledLeafEntry.
+func UnmarshalLeafEntry(entry *LeafEntry, ret *UnmarshaledLeafEntry) error {
+	if rest, err := tls.Unmarshal(entry.LeafInput, &ret.Leaf); err != nil {
+		return fmt.Errorf("failed to unmarshal MerkleTreeLeaf: %v", err)
 	} else if len(rest) > 0 {
-		return nil, fmt.Errorf("trailing data (%d bytes) after MerkleTreeLeaf for index %d", len(rest), index)
+		return fmt.Errorf("MerkleTreeLeaf: trailing data %d bytes", len(rest))
 	}
 
-	var err error
-	entry := LogEntry{Index: index, Leaf: leaf}
-	switch leaf.TimestampedEntry.EntryType {
+	switch eType := ret.Leaf.TimestampedEntry.EntryType; eType {
 	case X509LogEntryType:
+		ret.IsPrecert = false
 		var certChain CertificateChain
-		if rest, err := tls.Unmarshal(leafEntry.ExtraData, &certChain); err != nil {
-			return nil, fmt.Errorf("failed to unmarshal ExtraData for index %d: %v", index, err)
+		if rest, err := tls.Unmarshal(entry.ExtraData, &certChain); err != nil {
+			return fmt.Errorf("failed to unmarshal CertificateChain: %v", err)
 		} else if len(rest) > 0 {
-			return nil, fmt.Errorf("trailing data (%d bytes) after CertificateChain for index %d", len(rest), index)
-		}
-		entry.Chain = certChain.Entries
-		entry.X509Cert, err = leaf.X509Certificate()
-		if x509.IsFatal(err) {
-			return nil, fmt.Errorf("failed to parse certificate in MerkleTreeLeaf for index %d: %v", index, err)
+			return fmt.Errorf("CertificateChain: trailing data %d bytes", len(rest))
 		}
+		ret.Cert = *ret.Leaf.TimestampedEntry.X509Entry
+		ret.Chain = certChain.Entries
 
 	case PrecertLogEntryType:
+		ret.IsPrecert = true
 		var precertChain PrecertChainEntry
-		if rest, err := tls.Unmarshal(leafEntry.ExtraData, &precertChain); err != nil {
-			return nil, fmt.Errorf("failed to unmarshal PrecertChainEntry for index %d: %v", index, err)
+		if rest, err := tls.Unmarshal(entry.ExtraData, &precertChain); err != nil {
+			return fmt.Errorf("failed to unmarshal PrecertChainEntry: %v", err)
 		} else if len(rest) > 0 {
-			return nil, fmt.Errorf("trailing data (%d bytes) after PrecertChainEntry for index %d", len(rest), index)
+			return fmt.Errorf("PrecertChainEntry: trailing data %d bytes", len(rest))
 		}
-		entry.Chain = precertChain.CertificateChain
+		ret.Cert = precertChain.PreCertificate
+		ret.Chain = precertChain.CertificateChain
+
+	default:
+		return fmt.Errorf("unknown entry type: %v", eType)
```

Comment:
RFC 6962 implies that unknown types should be accepted by clients and treated as opaque (section 4.6). A mirror is both a log client and server so we should consider carefully how this works. 

@pphaneuf does the C++ mirror implementation parse log entries or treat them as blobs? How does it interact with the register cert flow parsing?

Reply:
That is a good point. This PR leaves handling unknown entry types behavior intact. I left a TODO to revisit this behavior.

One interesting question we need to answer before we make this "not an error" is what Trillian `LeafIdentityHash` we should assign to opaque blobs like this. I guess, we could just make it equal to `MerkleLeafHash` by default?